lib/src/css_mirroring/bidi_css_gen.dart

```diff
@@ -0,0 +1,355 @@
+///     3) Flipped Orientation specific: It is made from flipped css string.
+///                              Direction independent parts will be removed from it to keep only direction dependent parts of original css.
+///               eg: Initially contains foo { color: red; margin-right: 10px;} and will get modified to :host-context([dir="rtl"]) foo { margin-right: 10px;}
+///
+/// So for each of these transactions we extract toplevels of the originalCss and flippedCss. It iterates over these topLevels.
+/// If it is of type rule set
+///     Iterates over declarations in them.
+///     Depending on the mode of execution which could be [keepBidiNeutral], [keepOriginalBidiSpecific], [keepFlippedBidiSpecific],
+///     check if the declaration is to be removed and store their start and end points.
+///     Now if only some declarations have to be removed, remove them using their start and end points already stored.
+///     And if all declarations in a ruleset are to be removed, Remove the ruleset (No need to keep empty rule)
+///
+/// If it is of type Media or Host Directive
+///   eg:
+///    @media screen and (min-width: 401px) {
+///                  foo { margin-left: 13px }             /// Media Directive containing ruleset foo
+///                  }
+///   Directive --> RuleSets --> Declarations
+///   Pick a ruleset:
+///     stores removable declarations in it ->
+///     If only some of the declaration have to be removed -> remove them from transaction.
+///     If all declarations in ruleset removable -> store start and end of rule set(dont edit transaction because if all rulesets of directive have to be deleted then we will delete directive itself)
+///   If only some rulesets in Directive have to be removed -> remove them using store start and end points
+///   If all the rulesets have to be removed -> remove the Directive itself.
+///
+/// If it is a Direction Independent Directive
+///   eg:
+///     @charset "UTF-8";                                 /// Charset Directive
+///     @namespace url(http://www.w3.org/1999/xhtml);     /// Namespace Directive
+///  Keep it in one of the transaction and remove it from other two (Here we are keeping it in Orientation Neutral transaction).
+///
+/// We then combine these transactions to get the expected output css.
+
+typedef Future<String> CssFlipper(String src);
+
+class _PendingRemovals {
+  final String source;
+  final List<TreeNode> _topLevels;
+  final TextEditTransaction _transaction;
+  // List to contain start and end location of declarations in transaction.
+  var _declStartPoints = <int>[];
+  var _declEndPoints = <int>[];
+
+  _PendingRemovals(TextEditTransaction trans, this._topLevels)
+      : _transaction = trans,
+        source = trans.file.getText(0);
+
+  void addDeclRemoval(int start, int end) {
+    _declStartPoints.add(start);
+    _declEndPoints.add(end);
+  }
+
+  void commitDeclarations() {
+    for (int iDecl = 0; iDecl < _declStartPoints.length; iDecl++)
```

Comment:
https://www.dartlang.org/effective-dart/style/#do-use-curly-braces-for-all-flow-control-structures